File: iosApp/iosApp/Permissions/SystemPermissionHandlers/SystemPermissionHandler.swift
import ComposeApp

protocol SystemPermissionHandler {
    var id: String {get}
    func resolvePermissionState() -> PermissionStateType
    func canHandle(simplePlatformPermission: PlatformPermissionTypePlatformSinglePermission) -> Bool
    func requestPermission(onRequest: @escaping (_ permissionState: PermissionStateType) -> Void)
}

final class SystemPermissionHandlerWrapper: Hashable {
    let handler: any SystemPermissionHandler
    
    init(handler: any SystemPermissionHandler) {
        self.handler = handler
    }
    static func == (lhs: SystemPermissionHandlerWrapper, rhs: SystemPermissionHandlerWrapper) -> Bool {
        return lhs.handler.id == rhs.handler.id
    }
        
    func hash(into hasher: inout Hasher) {
        hasher.combine(handler.id)
    }
}

File: composeApp/src/iosMain/kotlin/org/example/kmpsamples/presentation/permissions/systemPermissionHandlers/SystemPermissionHandler.kt
package org.example.kmpsamples.permissions.systemPermissionHandlers

import org.example.kmpsamples.presentation.permissions.dtos.PermissionStateType
import org.example.kmpsamples.presentation.permissions.dtos.PlatformPermissionType.PlatformSinglePermission


interface SystemPermissionHandler {
    fun resolvePermissionState(): PermissionStateType
    fun canHandle(simplePlatformPermission: PlatformSinglePermission): Boolean
    fun requestPermission(onRequest: (permissionState: PermissionStateType) -> Unit)
}